test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java

```diff
@@ -940,6 +941,37 @@ public void waitNoPendingTasksOnAll() throws Exception {
         assertNoTimeout(clusterAdmin().prepareHealth(TEST_REQUEST_TIMEOUT).setWaitForEvents(Priority.LANGUID).get());
     }
 
+    /**
+     * Waits for the node {@code viaNode} to see {@code masterNodeName} as the master node in the cluster state and asserts that.
+     * Note that this does not guarantee that all other nodes in the cluster are on the same cluster state version already.
+     *
+     * @param viaNode the node to check the cluster state one
+     * @param masterNodeName the master node name that we wait for
+     */
+    public void awaitAndAssertMasterNode(String viaNode, String masterNodeName) throws Exception {
+        awaitClusterState(
```

Comment:
I'd be tempted to use `addTemporaryStateListener` here too, because (a) it is more harmonious with `awaitAndAssertMasterNotFound` and (b) it avoids the `throws Exception`.

Reply:
Yeah I thought about that too. I'm fine with using `addTemporaryStateListener` here too.

Is there a reason we would intentionally want the behavior of `awaitClusterState`? I've been tempted to make `awaitClusterState` just use `addTemporaryClusterState` (or even getting rid of `awaitClusterState`, but that's a larger change).

Reply:
> Is there a reason we would intentionally want the behavior of `awaitClusterState`?

IMO not really no. `ClusterStateObserver` is really old and suspiciously complicated given what it's actually doing. I'd be happy to migrate things away from it.